lib/src/util/dart_type_utilities.dart

```diff
@@ -47,13 +47,33 @@ class DartTypeUtilities {
   static bool hasInheritedMethod(MethodDeclaration node) =>
       lookUpInheritedMethod(node) != null;
 
+  static InterfaceType implementedInterface(
+      DartType type, String interface, String library) {
+    if (type is! InterfaceType) {
+      return null;
+    }
+    bool predicate(InterfaceType i) => isInterface(i, interface, library);
+
+    ClassElement element = type.element;
+    if (predicate(type)) {
+      return type;
+    }
+    if (!element.isSynthetic && type is InterfaceType) {
+      final interfaces = element.allSupertypes.where(predicate);
```

Comment:
You've definitely got the right approach. Unfortunately, this duplicates a lot of existing code. In the analyzer's model, an `InterfaceType` already represents the same thing I think you're trying to do with `_ClassElementBox`: the mapping from type parameters to type arguments. So, for example, when you have a type representing `List<int>`, what you really have is an object that knows that the type is defined by the element for `List<E>` (the element keeps the type parameters) and whose type arguments are the type for `int`.

If you ask the element for `List<E>` for its interfaces, it can tell you that it implements `Iterable<E>` (where this `E` is the type parameter associated with `List`). But if you ask the type `List<int>` for it's interfaces it can propagate the substitution and tell you that it implements `Iterable<int>`.

Given that this is all implemented and well tested (and will be maintained) it seems wrong to re-implement it here.